decidim-core/app/controllers/concerns/decidim/locale_switcher.rb

```diff
@@ -8,20 +8,20 @@ module LocaleSwitcher
     extend ActiveSupport::Concern
 
     included do
-      before_action :set_locale
+      around_action :switch_locale
       helper_method :current_locale, :available_locales, :default_locale
 
       # Sets the locale for the current session.
       #
       # Returns nothing.
-      def set_locale
+      def switch_locale(&action)
         locale = if params["locale"]
                    params["locale"]
                  elsif current_user && current_user.locale.present?
                    current_user.locale
                  end
 
-        I18n.locale = available_locales.include?(locale) ? locale : default_locale
+        I18n.with_locale(available_locales.include?(locale) ? locale : default_locale, &action)
```

Comment:
In case someone checks this code and wonders (like me) where are we changing the user locale, we currently do this here:

https://github.com/decidim/decidim/blob/f8ba6bd4320c4532dc5136b6cbb6e88ed78a9a88/decidim-core/app/controllers/decidim/locales_controller.rb#L8-L13